concepts/recursion/about.md

```diff
@@ -0,0 +1,166 @@
+    ordinal -- which weekday of the month, e.g. 2 (for the second)
+    """
+    output = []
+
+    for month in range(1, 13):
+        for day_num in range(1, 8):
+            if date(year, month, day_num).isoweekday() == weekday:
+                output.append(date(year, month, day_num + (ordinal - 1) * 7))
+                break
+    return output
+
+# find the second Wednesday of the month for all the months in 2022
+print(paydates_for_year(2022, 3, 2))
+
+```
+
+The program works, but Adya wonders if she can refactor the code to use fewer lines with less nested looping.
+She would also like to use less mutating of variables, such as `output`, `month`, and `day_num`, because she's read that it is good to minimize mutating state.
+She's heard about recursion and thinks about how to change her program to use a recursive approach.
+The variables that are created and mutated in her looping function will now be passed in as arguments.
+Instead of mutating the variables inside her function, she will pass updated values as arguments to the next call of her function.
+With those intentions she arrives at this recursive approach
+
+```python
+from datetime import date
+
+
+
+def paydates_for_year_rec(year, weekday, ordinal, month, day_num, output):
+    """Returns a list of the matching weekday dates
+    
+    Keyword arguments:
+    year    -- the year, e.g. 2022
+    weekday -- the weekday, e.g. 3 (for Wednesday)
+    ordinal -- which weekday of the month, e.g. 2 (for the second)
+    month   -- the month currently being processed
+    day_num -- the day of the month currently being processed
+    output  -- the list to be returned
+    """
+    if month == 13:
+        return output
+    if date(year, month, day_num).isoweekday() == weekday:
+        return paydates_for_year_rec(year, weekday, ordinal, month + 1, 1, output
+                                     + [date(year, month, day_num + (ordinal - 1) * 7)])
+    return paydates_for_year_rec(year, weekday, ordinal, month, day_num + 1, output)
+
+    # find the second Wednesday of the month for all the months in 2022
+    print(paydates_for_year_rec(2022, 3, 2, 1, 1, []))
+    
+```
+
+Adya is happy that there are no more nested loops, no mutated state, and 2 fewer lines of code than her looping solution.
+Later, when she learns about tuples,  Adya may consider other approaches, such as using a list comprehension with `Calendar.itermonthdates`.
+She may also become concerned that the recursive approach uses more steps than the looping approach, and so is less performant.
+But, when faced with a solution that involves a lot of nested looping, extensive state mutation, and complex conditional logic, Adya may try to work it out recursively first.
+She may find it easier to initially reason through the problem when nesting, mutation, and complexity are minimized.
+After she's worked out the basic logic she can then work at optimizing the recursive solution into a more performant looping approach.
+
```

Comment:
I like this example, but it feels in tone to me more like a concept exercise rather than an illustration of loops vs recursion.  My though here it that maybe we use the `collatz conjecture` example for code on  `loops` vs `recursion`, and we save **this** narrative for a recursion _concept exercise_.  

I don't think we lose anything by using `Collatz` here -- students will likely have tackled  `Collatz` when they do the numbers concept.  And that might work even better since they will be familiar with the code, and might then be able to reason about the differences better.

Another option is to use the code from this example, but remove/alter the narrative style.  I think the narrative is _good_ -- but its story-telling doesn't exactly mesh (for me) with the rest of the doc, which reads more "friendly/casual college lecture".

Reply:
I'm not sure collatz is as good an example for "why recursion", as it doesn't have the nested looping. But if the purpose is to just show the difference between the two outcomes, as opposed to the difference in the two thought processes, then I suppose collatz would be a straightforward, though somewhat dull, comparison.

I'm also not sure that the narrative style _needs_ to "mesh" in tone with the surrounding doc, as a story can serve a pedagogic purpose of drawing the student in to the recursive thought process in an engaging way. But if the story is off-putting, then yeah, it should go.